### ISSUE TITLE
HandBeliefに34牌種canonical axisのwait beliefを追加する

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong#82]
## 背景

現在の `HandBelief` は、1 wind の非公開手牌について、34基本牌種ごとの expected count と3色の赤5 probability を canonical な fixed-point representation として保持している。

今後、他家手牌推定を聴牌推定・待ち牌推定・待ち形推定・放銃危険度推定・heuristic / learned HandBeliefへ発展させるには、**「どの牌が構造上の待ちか」と「どの待ちmechanismによるものか」**を HandBelief 上で表現できる共通形式が欲しい。

当初は singleton / pair / adjacent taatsu / gapped taatsu といった未完成shapeを表現する案も検討したが、未聴牌時には手牌分解が一意に定まらないケースがあり、完全情報から教師ラベルを生成する際に恣意的な decomposition rule が必要になる。

一方、待ちに限定すれば、完全情報から deterministic な ground truth を作りやすい。

- 非聴牌: 全 wait table が 0
- 聴牌: 実際の structural wait と成立し得る wait mechanism を 0/1 で表現

将来の教師あり学習を考え、**完全情報から一意な multi-label ground truth を生成できること**を優先し、HandBeliefへ canonical wait belief representation を導入する。

## Goal

`HandBelief` に、既存の34基本牌種 canonical axis と同じ形を持つ wait probability table を追加する。

概念上は以下とする。

```text
HandBelief
├── expected_count_raw                  [34]
├── red_five_probability_raw             [3]
│
├── wait_probability_raw                [34] | None
│
├── tanki_wait_probability_raw          [34] | None
├── shanpon_wait_probability_raw        [34] | None
├── kanchan_wait_probability_raw        [34] | None
├── penchan_wait_probability_raw        [34] | None
├── ryanmen_low_side_probability_raw    [34] | None
├── ryanmen_high_side_probability_raw   [34] | None
└── kokushi_wait_probability_raw        [34] | None
```

すべての wait table は canonical `tile_type_index` 順の length 34 とし、既存 HandBelief と同じ tile axis を利用する。

## Design principle: すべて34牌種axisに揃える

各 table は待ち形ごとに必要な slot 数へ圧縮せず、**すべて length 34** とする。

国士待ちも13牌種だけの専用axisにはせず、`kokushi_wait_probability_raw[34]` として保持し、成立し得ない slot は canonical zero とする。

これにより、将来も

```text
wind × feature channel × tile type
```

という一貫した構造を維持し、tensor化・serialization・comparison・calibration・learned model の target / output へ同じ tile axis を使えるようにする。

## Structural wait の定義

このIssueでいう `wait` は **legal winning action や yaku 付き和了可否ではなく、structural completion wait** を意味する。

`wait_probability[t]` は、

> 現在の手牌構造に牌種 `t` を1枚加えたとき、通常手・七対子・国士無双のいずれかの完成形を構成できる probability

を表す。

したがって以下とは分離する。

- furiten
- ron / tsumo action legality
- yaku の有無
- 点数状況
- riichi状態
- Policy action legality
- 現時点でその牌が山等に残っているか

役なし副露手でも、形として完成する牌なら structural wait になり得る。

また、場に同種牌が4枚見えていて remaining copies が0であっても、手牌構造上の待ちであるなら ground truth の `wait[t]` は1になり得る。remaining tile availability は別feature / beliefとして扱う。

ただし、物理的に5枚目の同一牌を必要とするなど、麻雀牌の物理制約上成立しないcandidateはwaitとしない。

## Primary table: `wait_probability`

```text
wait_probability[t]
```

は、牌種 `t` が structural completion wait である probability を表す。

完全情報 ground truth では binary とする。

- 非聴牌: 全34slotを0
- 聴牌: structural wait となる牌種を1
- 多面待ち: 複数slotが同時に1

したがって table sum に 1.0 制約は設けない。

`wait_probability[34]` を primary target とし、以下の mechanism table は auxiliary / explanatory target と位置付ける。

## Wait mechanism tables

各 mechanism table は、

> 牌種 `t` が、その mechanism によって structural hand completion を成立させる probability

を表す。

完全情報 ground truth では0/1のmulti-labelとする。

### Tanki

```text
tanki_wait_probability[t]
```

1枚の singleton が同種牌 `t` を得ることで completion する場合を表す。

七対子聴牌も tanki に包含する。

例:

```text
11m 22m 33p 44p 55s 66s 東
```

なら

```text
wait[東]  = 1
tanki[東] = 1
```

とする。

ここで `tanki` は hand family ではなく wait mechanism である。通常手単騎と七対子を区別する必要が生じた場合は、別途 hand-family feature を追加する。

国士待ちは tanki へ包含せず、専用 `kokushi` channel で表現する。

### Shanpon

```text
shanpon_wait_probability[t]
```

牌種 `t` を得ることで対子が刻子となり、双碰待ちを完成する場合を表す。

全34牌種を valid key とする。

### Kanchan

```text
kanchan_wait_probability[t]
```

牌種 `t` が嵌張の中央を埋める completion tile である場合を表す。

例:

```text
2m4m -> 3m
```

なら `kanchan[3m] = 1`。

各色 rank 2..8 のみ non-zero になり得る。それ以外は canonical zero とする。

### Penchan

```text
penchan_wait_probability[t]
```

辺張を完成する牌を表す。

```text
1m2m -> 3m
8m9m -> 7m
```

各色 rank 3 / 7 のみ non-zero になり得る。それ以外は canonical zero とする。

### Ryanmen low side

```text
ryanmen_low_side_probability[t]
```

待ち牌 `t` が、元の両面taatsuより**数値の低い側**を完成する場合を表す。

例:

```text
2m3m -> 1m
```

なら `ryanmen_low_side[1m] = 1`。

各色 rank 1..6 のみ non-zero になり得る。

### Ryanmen high side

```text
ryanmen_high_side_probability[t]
```

待ち牌 `t` が、元の両面taatsuより**数値の高い側**を完成する場合を表す。

例:

```text
2m3m -> 4m
```

なら `ryanmen_high_side[4m] = 1`。

各色 rank 4..9 のみ non-zero になり得る。

`low_side` / `high_side` は待ち牌自身のrankが元taatsuより低いか高いかを意味し、元taatsuのindexやsequence startを意味しない。

### Kokushi

```text
kokushi_wait_probability[t]
```

牌種 `t` が国士無双を完成する structural wait である場合を表す。

validなのは13幺九牌のみ。

```text
1m 9m
1p 9p
1s 9s
東 南 西 北 白 發 中
```

その他21slotは canonical zero とする。

- 国士13面待ち: 13slotすべて1
- 欠け1種待ち: 該当1slotのみ1

## Deterministic multi-label ground truth contract

完全情報からmechanism labelを生成するとき、**正解となるdecompositionを1つだけ選ばない**。

candidate tile `t` について、`t` を加えることで成立するすべての構造上有効な完成 decomposition / hand family を考慮する。

そのうち少なくとも1つで当該 wait mechanism が成立するなら、その mechanism table の `t` を1とする。

したがって、同一 tile について複数mechanismが同時に1でも合法である。

```text
wait[t] = 1
```

かつ、例えば

```text
tanki[t] = 1
ryanmen_high_side[t] = 1
```

のようなmulti-labelを許容する。

これにより、曖昧な完成手について恣意的なdecomposition choiceを教師ラベルへ持ち込まない。

### 完全情報 ground truth

真のconcealed handと既知のpublic meldが分かっている場合、各cellは binary 0/1 とする。

非聴牌は all-zero であり、unknownではない。

### Hidden-hand belief

実際の他家 `HandBelief` では hidden hand に対する推定値なので、各cellは

```text
0.0 <= probability <= 1.0
```

となる。

例:

```text
wait_probability[5m] = 0.42
```

は「相手が現在5mを structural wait としている確率を0.42と推定している」ことを意味する。

## Availability contract: 3 levels

既存 estimator は wait belief をまだ生成していないため、zero defaultを与えてはならない。

次の3状態を許容する。

### Level 0: wait belief未提供

```text
wait_probability = None
mechanism group  = all None
```

### Level 1: primary waitのみ提供

```text
wait_probability = [34]
mechanism group  = all None
```

将来、まず「何待ちか」だけを推定する軽量 estimator / learned model を実装できるようにする。

### Level 2: wait + mechanism一式を提供

```text
wait_probability = [34]
mechanism group  = all [34]
```

mechanism group 内のpartial availabilityは許可しない。

```text
kanchanあり / kokushiなし
```

のような状態は禁止する。

また、mechanism group が存在するのに `wait_probability` が `None` の状態も禁止する。

これにより、

```text
None     = estimatorがそのfeatureを提供していない
all-zero = estimatorが全wait probabilityを0と推定している
```

を区別する。

## Fixed-point representation

既存 HandBelief の fixed-point contract を再利用する。

```text
SCALE = 8192 = 2^13
semantic probability : 0.0 .. 1.0
raw probability      : 0 .. SCALE
```

新しいscaleやfloat storageは導入しない。

semantic -> raw 変換も既存の round-half-to-even contract に従う。

実装上、重複回避のため generic probability helper を導入してもよいが、このIssueでは既存public APIを破壊しない。

## Semantic accessors

通常のPolicy / estimator codeはraw値ではなくsemantic accessorを利用できるようにする。

例:

```python
hand_belief.wait_probability(tile_type)
hand_belief.tanki_wait_probability(tile_type)
hand_belief.shanpon_wait_probability(tile_type)
hand_belief.kanchan_wait_probability(tile_type)
hand_belief.penchan_wait_probability(tile_type)
hand_belief.ryanmen_low_side_probability(tile_type)
hand_belief.ryanmen_high_side_probability(tile_type)
hand_belief.kokushi_wait_probability(tile_type)
```

未提供featureについてunknownを保持できるAPIとする。

## Ground-truth consistency

完全情報のLevel 2 binary ground truthでは、各tile `t` について

```text
wait[t] = 1
```

iff 少なくとも1つのmechanism channelで`t = 1`とする。

一方、probabilisticな hidden-hand belief では各値はmarginal probabilityであり、mechanism event間の相関を保持していない。

したがって `wait_probability` をmechanism probabilityの単純な sum / max / OR formula として定義せず、constructorでも厳密な代数制約を課さない。

これらのmarginal値から単一のjoint distributionを復元できることも保証しない。

## Public meldとの関係

waitはconcealed tilesだけでなく公開副露にも依存する。

そのため、この追加により `HandBelief` は牌種marginalだけでなく、

> concealed hand とそのwindについて観測可能なpublic hand stateを条件とした derived hand-state belief

も保持できる型へ拡張される。

完全情報ground truthを将来生成する際も、concealed tilesだけでなく既知のmeldを含めてstructural waitを判定する。

## Scope

- `HandBelief` に canonical wait belief representation を追加する
- すべてのwait channelを既存と同じ34牌種canonical axisへ統一する
- `wait` / `tanki` / `shanpon` / `kanchan` / `penchan` を定義する
- ryanmenの2筋を low-side / high-side として区別する
- 七対子待ちを tanki へ包含する
- 国士専用wait channelを追加する
- invalid slotの canonical zero contract を固定する
- `[0,1]` fixed-point probability contract を固定する
- Level 0 / 1 / 2 availability contractを実装する
- `None != all-zero` を維持する
- semantic accessorを追加する
- constructor validationを追加する
- 必要なpublic exportを更新する
- unit testsを追加する
- docs / docstringを同期する

## Non-goals

このIssueでは以下を実装しない。

- hidden handからwait probabilityを推定するheuristic
- conditional-uniform estimatorへのwait推定追加
- learned estimator / neural network
- training dataset生成
- 完全情報から0/1 wait ground truthを生成するbuilder / decomposition enumerator
- Policyへのwait belief利用
- 放銃危険度計算
- furiten判定との統合
- yaku / 打点推定
- legal ron / tsumo probability
- score / rank utility
- remaining tile availabilityをwait probabilityへ掛け込む処理
- ArenaへのAI内部belief公開
- GameTraceへのwait belief追加
- 通常手 / 七対子 / 国士そのもののhand-family probability推定

完全情報ground-truth builderは、このcanonical representationを先に固定した後の後続Issueとする。

## Tests

少なくとも以下を検証する。

1. Level 0: wait / mechanismすべて未提供を表現できる
2. Level 1: waitのみ提供できる
3. Level 2: wait + mechanism一式を提供できる
4. mechanism groupのpartial availabilityを拒否する
5. wait未提供でmechanismだけ提供する状態を拒否する
6. 各tableがlength 34を要求する
7. raw値 `0..SCALE` のみ受理する
8. range外・`bool`・non-intをfail-closedで拒否する
9. all-zero tableが合法である
10. table sumが1.0を超えても合法である
11. `wait` は34slotすべてvalid
12. `tanki` は34slotすべてvalid
13. `shanpon` は34slotすべてvalid
14. `kanchan` は各色rank 2..8のみnon-zero可能
15. `penchan` は各色rank 3 / 7のみnon-zero可能
16. `ryanmen low-side` は各色rank 1..6のみnon-zero可能
17. `ryanmen high-side` は各色rank 4..9のみnon-zero可能
18. `kokushi` は13幺九牌のみnon-zero可能
19. invalid slotのnon-zeroを拒否する
20. 同一tileについて複数mechanismが同時にnon-zeroでも合法
21. semantic accessorがfixed-point値を正しく返す
22. 未提供featureとzero probabilityを区別できる
23. probabilistic beliefではwaitとmechanism間にsum/max等の過剰なconstraintを課さない
24. 既存 `HandBelief` constructor利用箇所を壊さない
25. 既存conditional-uniform / exact-self等のbehaviorを変更しない

## Acceptance criteria

- [ ] `HandBelief` が canonical wait belief を保持できる
- [ ] 全wait tableが既存と同じ34牌種canonical axisを使用する
- [ ] `wait_probability[34]` をprimary structural-wait beliefとして保持できる
- [ ] tanki / shanpon / kanchan / penchanを34長tableで表現できる
- [ ] ryanmenの2筋を low-side / high-side として区別できる
- [ ] 七対子の待ちを tanki へ包含する
- [ ] 国士待ちを専用34長channelで表現できる
- [ ] 国士のinvalid 21slotはcanonical zeroである
- [ ] structural waitとlegal winning action / yaku / furiten / remaining copiesを分離している
- [ ] 非聴牌ground truthをall-zeroとして表現できる
- [ ] 完全情報では全decompositionを考慮したdeterministic multi-label ground truthを定義できる
- [ ] wait mechanismはmulti-labelを許容する
- [ ] 既存 `SCALE = 8192` fixed-point representationを共有する
- [ ] Level 0 / 1 / 2 のfeature availabilityを表現できる
- [ ] mechanism groupはall-or-noneである
- [ ] `None = 未提供` と `all-zero = probability 0` を区別する
- [ ] hidden-hand probabilistic beliefに不適切なsum/max constraintを課さない
- [ ] 既存HandBelief APIとの後方互換性を維持する
- [ ] estimator / ground-truth builder / Policy / learning実装をこのIssueへ含めない
- [ ] unit tests / docs / docstringが同期する
- [ ] Ruff format / Ruff check / full test suiteが成功する

## 将来の接続

```text
完全情報の実手牌 + public meld
              |
              v
       exact structural wait ground truth
              |
      34牌種canonical axis
              |
   +----------+-----------+
   |                      |
   v                      v
wait[34]             mechanism channels
                         |
             +-----------+-----------+
             |      |      |         |
           tanki shanpon kanchan ... kokushi
                         |
                         v
                supervised learning
                         |
                         v
                    HandBelief
                         |
             +-----------+-----------+
             |                       |
             v                       v
      remaining tile belief     value / safety
             |                       |
             +-----------+-----------+
                         v
                 deal-in risk / defense
                         |
                         v
                    push / fold
```
